src/Grpc.AspNetCore.Server/Internal/ServerCallDeadlineManager.cs

```diff
@@ -0,0 +1,264 @@
+
+            return _deadlineExceededCompleteTcs.Task;
+        }
+
+        private static void DeadlineExceededCallback(object? state) => _ = ((ServerCallDeadlineManager)state!).DeadlineExceededAsync();
+
+        private static void DeadlineExceededLongCallback(object? state)
+        {
+            var (manager, maxTimerDueTime) = (ValueTuple<ServerCallDeadlineManager, long>)state!;
+            var remaining = manager.Deadline - manager._systemClock.UtcNow;
+            if (remaining <= TimeSpan.Zero)
+            {
+                _ = manager.DeadlineExceededAsync();
+            }
+            else
+            {
+                // Deadline has not been reached because timer maximum due time was smaller than deadline.
+                // Reschedule DeadlineExceeded again until deadline has been exceeded.
+                GrpcServerLog.DeadlineTimerRescheduled(manager._serverCallContext.Logger, remaining);
+
+                manager._longDeadlineTimer.Change(manager.GetTimerDueTime(remaining, maxTimerDueTime), Timeout.Infinite);
+            }
+        }
+
+        private void RequestAborted()
+        {
+            // Call is complete if the request has aborted
+            lock (this)
+            {
+                IsCallComplete = true;
+                _deadlineCts!.Cancel();
+            }
+        }
+
+        private long GetTimerDueTime(TimeSpan timeout, long maxTimerDueTime)
+        {
+            // Timer has a maximum allowed due time.
+            // The called method will rechedule the timer if the deadline time has not passed.
+            var dueTimeMilliseconds = timeout.Ticks / TimeSpan.TicksPerMillisecond;
+            dueTimeMilliseconds = Math.Min(dueTimeMilliseconds, maxTimerDueTime);
+            // Timer can't have a negative due time
+            dueTimeMilliseconds = Math.Max(dueTimeMilliseconds, 0);
+
+            return dueTimeMilliseconds;
+        }
+
+        public void SetCallEnded()
+        {
+            lock (this)
+            {
+                IsCallComplete = true;
+            }
+        }
+
+        public bool TrySetCallComplete()
+        {
+            lock (this)
+            {
+                if (!IsDeadlineExceededStarted)
+                {
+                    IsCallComplete = true;
+                    return true;
+                }
+
+                return false;
+            }
+        }
+
+        private async Task DeadlineExceededAsync()
+        {
+            if (!TryStartExceededDeadline())
+            {
+                return;
+            }
+
+            try
+            {
+                await _serverCallContext.DeadlineExceededAsync();
+                lock (this)
+                {
+                    IsCallComplete = true;
+
+                    // Canceling CTS will trigger registered callbacks.
+                    // Exception could be thrown from them.
+                    _deadlineCts?.Cancel();
```

Comment:
Since cancelling _deadlineCts can invoke user-defined callbacks, it would feel safer if all the attempts to cancel it were done outside of a lock (and better yet, dispatched to the ThreadPool), but I don't know the codebase well enough to know if this is really necessary.